src/storage/src/hummock/compactor/shared_buffer_compact.rs

```diff
@@ -476,6 +403,92 @@ pub async fn merge_imms_in_memory(
     })
 }
 
+fn generate_splits(
+    payload: &UploadTaskPayload,
+    existing_table_ids: &HashSet<u32>,
+    storage_opts: &StorageOpts,
+) -> (Vec<KeyRange>, u64, u32) {
+    let mut size_and_start_user_keys = vec![];
+    let mut compact_data_size = 0;
+    for imm in payload {
+        let data_size = {
+            // calculate encoded bytes of key var length
+            (imm.kv_count() * 8 + imm.size()) as u64
+        };
+        compact_data_size += data_size;
+        size_and_start_user_keys.push((data_size, imm.start_user_key()));
+    }
+    size_and_start_user_keys.sort_by(|a, b| a.1.cmp(&b.1));
+    let mut splits = Vec::with_capacity(size_and_start_user_keys.len());
+    splits.push(KeyRange::new(Bytes::new(), Bytes::new()));
+    let mut key_split_append = |key_before_last: &Bytes| {
+        splits.last_mut().unwrap().right = key_before_last.clone();
+        splits.push(KeyRange::new(key_before_last.clone(), Bytes::new()));
+    };
+    let sstable_size = (storage_opts.sstable_size_mb as u64) << 20;
+    let parallel_compact_size = (storage_opts.parallel_compact_size_mb as u64) << 20;
+    let parallelism = std::cmp::min(
+        storage_opts.share_buffers_sync_parallelism as u64,
+        size_and_start_user_keys.len() as u64,
+    );
+    let sub_compaction_data_size = if compact_data_size > parallel_compact_size && parallelism > 1 {
+        compact_data_size / parallelism
+    } else {
+        compact_data_size
+    };
+    // mul 1.2 for other extra memory usage.
+    let mut sub_compaction_sstable_size =
+        std::cmp::min(sstable_size, sub_compaction_data_size * 6 / 5);
+    let mut split_weight_by_vnode = 0;
+    if existing_table_ids.len() > 1 {
+        if parallelism > 1 && compact_data_size > sstable_size {
```

Comment:
I think we can move this check before `size_and_start_user_keys.sort_by` to avoid unnecessary sorting given that parallelism is 1 in most cases.

Reply:
= = But this array will not be large, and it only takes a little more time to sort it than construct it.